File: product-consumer-service/src/main/java/de/hdi/product_consumer_service/controller/OrderController.java
package de.hdi.product_consumer_service.controller;

import java.util.List;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import de.hdi.product_consumer_service.model.Order;
import de.hdi.product_consumer_service.model.OrderItem;
import de.hdi.product_consumer_service.model.Product;
import de.hdi.product_consumer_service.service.OrderService;
import de.hdi.product_consumer_service.service.ProductClientService;
import lombok.RequiredArgsConstructor;

@RestController
@RequestMapping("/api/orders")
@RequiredArgsConstructor
public class OrderController {

    private final OrderService orderService;
    private final ProductClientService productClientService;

    @GetMapping
    public ResponseEntity<List<Order>> getAllOrders() {
        return ResponseEntity.ok(orderService.getAllOrders());
    }

    @GetMapping("/{id}")
    public ResponseEntity<Order> getOrderById(@PathVariable Integer id) {
        return orderService.getOrderById(id)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @PostMapping
    public ResponseEntity<Order> createOrder(@RequestBody Order order) {
        Order createdOrder = orderService.createOrder(order);
        return ResponseEntity.status(HttpStatus.CREATED).body(createdOrder);
    }

    @PutMapping("/{id}")
    public ResponseEntity<Order> updateOrder(@PathVariable Integer id, @RequestBody Order order) {
        return orderService.updateOrder(id, order)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteOrder(@PathVariable Integer id) {
        if (orderService.deleteOrder(id)) {
            return ResponseEntity.noContent().build();
        }
        return ResponseEntity.notFound().build();
    }

    @PostMapping("/{orderId}/items")
    public ResponseEntity<Order> addItemToOrder(@PathVariable Integer orderId, @RequestBody OrderItem item) {
        return orderService.addItemToOrder(orderId, item)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @DeleteMapping("/{orderId}/items/{itemId}")
    public ResponseEntity<Order> removeItemFromOrder(@PathVariable Integer orderId, @PathVariable Integer itemId) {
        return orderService.removeItemFromOrder(orderId, itemId)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @GetMapping("/products")
    public ResponseEntity<List<Product>> getAllProducts() {
        List<Product> products = productClientService.getAllProducts();
        return ResponseEntity.ok(products);
    }

    @GetMapping("/products/{id}")
    public ResponseEntity<Product> getProductById(@PathVariable Integer id) {
        try {
            Product product = productClientService.getProduct(id);
            return ResponseEntity.ok(product);
        } catch (Exception e) {
            return ResponseEntity.notFound().build();
        }
    }

    @PostMapping("/products")
    public ResponseEntity<Product> createProduct(@RequestBody Product product) {
        Product createdProduct = productClientService.createProduct(
                product.getName(), product.getDescription(), product.getPrice());
        return ResponseEntity.status(HttpStatus.CREATED).body(createdProduct);
    }
}

File: order-kotlin-service/src/main/kotlin/de/hdi/order_kotlin_service/web/OrderController.kt
package de.hdi.order_kotlin_service.web

import de.hdi.order_kotlin_service.model.OrderRequest
import de.hdi.order_kotlin_service.model.OrderResponse
import de.hdi.order_kotlin_service.model.OrderStatus
import de.hdi.order_kotlin_service.service.OrderService
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.time.LocalDateTime

data class OrderResponseDTO(
    val id: String,
    val productId: Int,
    val productName: String,
    val quantity: Int,
    val totalPrice: Double,
    val customerEmail: String,
    val createdAt: LocalDateTime,
    val status: OrderStatus
)

@RestController
@RequestMapping("/api/orders")
class OrderController(private val orderService: OrderService) {

    @PostMapping
    fun createOrder(@RequestBody orderRequest: OrderRequest): ResponseEntity<OrderResponseDTO> {
        val response = orderService.createOrder(orderRequest)
        return if (response != null) {
            ResponseEntity.status(HttpStatus.CREATED).body(
                responseBody(response)
            )
        } else {
            ResponseEntity.badRequest().build()
        }
    }

    @GetMapping("/{orderId}")
    fun getOrder(@PathVariable orderId: String): ResponseEntity<OrderResponseDTO> {
        val response = orderService.getOrder(orderId)
        return if (response != null) {
            ResponseEntity.ok(responseBody(response))
        } else {
            ResponseEntity.notFound().build()
        }
    }

    @GetMapping
    fun getAllOrders(): ResponseEntity<List<OrderResponseDTO>> {
        val orders = orderService.getAllOrders()
        val orderDTOs = orders.map {
            responseDto(it)
        }
        return ResponseEntity.ok(orderDTOs)
    }

    private fun responseDto(it: OrderResponse) = responseBody(it)

    @PatchMapping("/{orderId}/status")
    fun updateOrderStatus(
        @PathVariable orderId: String,
        @RequestParam status: OrderStatus
    ): ResponseEntity<OrderResponseDTO> {
        val response = orderService.updateOrderStatus(orderId, status)
        return if (response != null) {
            ResponseEntity.ok(responseBody(response))
        } else {
            ResponseEntity.notFound().build()
        }
    }

    @DeleteMapping("/{orderId}")
    fun deleteOrder(@PathVariable orderId: String): ResponseEntity<Void> {
        val deleted = orderService.deleteOrder(orderId)
        return if (deleted) {
            ResponseEntity.noContent().build()
        } else {
            ResponseEntity.notFound().build()
        }
    }

    private fun responseBody(response: OrderResponse) = OrderResponseDTO(
        id = response.id,
        productId = response.productId,
        productName = response.productName,
        quantity = response.quantity,
        totalPrice = response.totalPrice,
        customerEmail = response.customerEmail,
        createdAt = response.createdAt,
        status = response.status
    )
}